astropy/timeseries/downsample.py

```diff
@@ -13,13 +13,36 @@
 __all__ = ["aggregate_downsample"]
 
 
+def nanmean_reduceat(data, indices):
+    mask = np.isnan(data)
+
+    if mask.any():  # If there are NaNs
+        # Create a writeable copy and mask NaNs
+        data_copy = data.copy()
+        data_copy[mask] = 0
+        count_data = np.add.reduceat(~mask, indices)
+        count_data = count_data.astype(float)
+        count_data[count_data == 0] = np.nan
+    else:
+        data_copy = data
+        # Derive counts from indices
+        count_data = np.diff(indices, append=len(data))
+        count_data[count_data <= 0] = 1
+
+    sum_data = np.add.reduceat(data_copy, indices)
```

Comment:
I suggest we avoid naming a variable `data_copy` if it's only a copy *sometimes*. Instead, let's shadow the initial variable as needed.

```suggestion
        data = data.copy()
        data[mask] = 0
        count_data = np.add.reduceat(~mask, indices)
        count_data = count_data.astype(float)
        count_data[count_data == 0] = np.nan
    else:
        # Derive counts from indices
        count_data = np.diff(indices, append=len(data))
        count_data[count_data <= 0] = 1

    sum_data = np.add.reduceat(data, indices)
```